### DISCUSSION TITLE
Enabling the I2C controller and creating a bus

### DISCUSSION THREAD

**Discussion by skyne98** [easytarget/MQ-Pro-IO#5] [GPIO]
Hi! Thanks for the unique and extremely useful resource! 

I am somewhat of an embedded linux noob, so I am trying to figure things out step by step with my `mq pro`. Currently I am trying to initialize i2c to communicate with the "Adafruit 128x64 OLED Bonnet"'s OLED panel with the SD1306 driver. 

Problem I am currently having is with the bus and controller. You wrote that the correct way of doing things is using the vanilla mq pro device tree and then using `gpiod` and `pinctl` to dynamically manage devices. Can you please help me out on how that works? 

All the resources I find tell me to compile a device tree to enable i2c and I cannot find any that mentions how to use the two to do it dynamically D:

Current goal: initialize an i2c bus on pins 3 and 5.

Thanks!

**Comment by easytarget** [comment]
I converted this to a discussion; seems the easiest way to discuss things. ;-)

<edited away my long answer, it no longer really applies, see below>

**Comment by easytarget** [comment]
@skyne98 
OK; So the guide was pretty light on details of GPIO use when this thread started.. But I hope it is much better now, I have done a few days experimenting and updating docs, and I just pushed all that work to main in PR #6
* (Plus some other updates to the custom tree tooling to make life easier when maintaining a machine)

The first thing to note is that the **Correct** solution for enabling hardware GPIO features is to use Device Tree Overlays. But I have not done this yet. So I still present a 'make a full custom device tree' solution.

I have I2C, PWM and general Digital IO covered there, it all works
* I have read temperatures from a BME280 sensor via I2C, and driven a LED via PWM.
* The only thing not tested is SPI, I don't have any test hardware for that, nor do I plan to use it in my projects.

There is now a GPIO guide at: [GPIO-examples.md](/easytarget/MQ-Pro-IO/GPIO-examples.md) with examples and notes.

I have also improved and expanded the alternate tree examples; adding one that closely mimics a standard RPi HAT layout.

**Reply by skyne98** [reply]
Thanks a lot! Will be really interested in the overlays solution! Thanks for updating the guide too, will check it out, just got my push in headers.

Currently, already applied your example SPI/I2C tree and will be trying it out with SD1306.

**Reply by easytarget** [reply]
Cool: Please let me know how it goes; I also want to drive a SD1306.
- Check out: https://github.com/easytarget/SBCEye

I wrote this a couple of years ago, it has a BME280 and a SD1306 on a RP3, with a web interface and GPIO relay control.
- It currently uses a Pi-Specific python libraries for sensor, display and GPIO; I want to convert it to use generic `lgpio` and get it running on the MQ Pro.
- Then I want to make it also send/recieve data via LoRa from a sensor on my sailboat, which is on a nearby lake.
- Long term plan.. might never happen

Also sort of related; I have done a lot of work with a SD1306 in my micropython font library: https://github.com/easytarget/microPyEZfonts

**Comment by easytarget** [comment]
@skyne98 
An update: I did a load of reading and some experimenting with a simple test overlay last night. I can compile it, but failed to deploy it. There is a bit more info in issue #7 

As I was doing this I realised that an overlay based solution has pretty much the same overhead as just compiling a full tree. They still need to be (ideally) recompiled for each new kernel, and cannot be applied 'dynamically'(*), so you still need to reboot after compiling and installing. 👎 

I might come back to this later; but for now please go with the 'full tree' method.

(*) There is a way.. but it involves recompiling a custom kernel for each version with an unsupported option. ☹️

**Reply by skyne98** [reply]
Thanks for the investigation! Will admit, a little surprised it offers so few benefits 🙃 I guess the main shtick is being able to enable parts of the device tree without needing to actually recompile anything, just use ready overlays?

**Reply by easytarget** [reply]
@skyne98 
Yes; for a device vendor like the Pi foundation it is great, they can distribute one base kernel+tree, and a bunch of corresponding overlays to let people modify things in a standard way. They can deliver both the kernel and overlays simultaneously.

For 'home use' it makes less sense, since you still need to wait for the kernel update to land before recompiling overlays, just the same as for full trees.

The ideal solution would be to create a 'mqpro-overlays' package that has standard overlays and install tooling. Then get that accepted and distributed along with the main (ubuntu) distro, or host in a seperate repo. Way too much work for me. ;-)

**Reply by skyne98** [reply]
I will solder the header today and see if i2c works as expected 😎

**Reply by easytarget** [reply]
Hi @skyne98 
Did you have any success? 

I did some more playing over the weekend and got a SSD1306 working on my demo system, then wrote a demo python script that pulls readings from the bme280 and displays them on the OLED. 

I've written it all up at:
https://github.com/easytarget/MQ-Pro-IO/blob/main/GPIO-examples.md

**Reply by skyne98** [reply]
I would assume you made 1306 via the GPIO pins? Do you mind sharing a repo? It's my first time soldering so I first tried it on my cheaper Longan Nano and had no success there yet with SD1351. Thinking about coming back to the MQ Pro x_x

**Reply by easytarget** [reply]
@skyne98  Yes, both the BME280 sensor and the SSD1306 are connected in parallel to header pins 3 and 5 (SDA and SCL respectively),  VCC goes to 3.3v and GND to gnd. (Do not connect them to the 5v supply, it will damage the cpu.) There is a picture of my test rig in the GPIO document.

The GPIO document shows how to install gpiod and the i2c-tools, then run i2cdetect to list all the devices on the bus. If the display is detected there you are correctly wired and configured.

You'll also need a custom Device Tree (the `lora-hat` tree in `alt-trees` is what I am testing with. Follow the instructions in the repo to build and install it. I wont be providing precompiled trees.

The demo program is written in Python, installing the requirements and running it are covered in the GPIO doc. I dont have examples using C or other languages.

**Reply by skyne98** [reply]
Thanks! Will post an update when I try it out :) I hope I didn't plug 5v already a couple of times, the board still works, some components did say they were rated for 5v, do those 5v damage something inside the board when passed to ground?
